File: android/src/main/kotlin/io/nativebridge/FlutterNativeBridgePlugin.kt
package io.nativebridge

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.lang.reflect.Method

/**
 * Flutter Native Bridge Plugin
 *
 * Bridges Flutter and native Android code using reflection.
 * Supports both MethodChannel (request-response) and EventChannel (streams).
 */
class FlutterNativeBridgePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var binaryMessenger: BinaryMessenger? = null
    private var activity: Activity? = null

    companion object {
        private const val TAG = "FlutterNativeBridge"
        private const val CHANNEL_NAME = "flutter_native_bridge"
        private const val EVENT_CHANNEL_PREFIX = "flutter_native_bridge/events/"
        private val registeredObjects = mutableMapOf<String, Any>()

        // Track active event channels and their sinks
        private val activeEventChannels = mutableMapOf<String, EventChannel>()
        private val activeStreamSinks = mutableMapOf<String, EventSinkWrapper>()

        // Reference to the plugin instance for setting up event channels
        private var pluginInstance: FlutterNativeBridgePlugin? = null

        /**
         * Register an object with a custom name.
         * Use this to avoid conflicts when classes have the same simple name.
         */
        @JvmStatic
        fun register(name: String, obj: Any) {
            if (registeredObjects.containsKey(name)) {
                val existing = registeredObjects[name]!!::class.java.name
                val new = obj::class.java.name
                android.util.Log.w(TAG, "Overwriting '$name': $existing -> $new")
            }
            registeredObjects[name] = obj
            // Setup event channels for newly registered object
            pluginInstance?.setupEventChannelsForObject(name, obj)
        }

        /**
         * Register an object using its class simple name.
         * Warns if a class with the same name is already registered.
         */
        @JvmStatic
        fun register(obj: Any) {
            val name = obj::class.java.simpleName
            if (registeredObjects.containsKey(name)) {
                val existing = registeredObjects[name]!!::class.java.name
                val new = obj::class.java.name
                if (existing != new) {
                    android.util.Log.e(TAG,
                        "Name conflict! '$name' already registered as $existing. " +
                        "Use register(\"CustomName\", obj) for $new")
                    return  // Don't overwrite with conflicting class
                }
            }
            registeredObjects[name] = obj
            // Setup event channels for newly registered object
            pluginInstance?.setupEventChannelsForObject(name, obj)
        }

        @JvmStatic
        fun unregister(name: String) {
            // Clean up any event channels for this object
            val prefix = "${EVENT_CHANNEL_PREFIX}$name."
            activeEventChannels.keys.filter { it.startsWith(prefix) }.forEach { channelName ->
                activeEventChannels[channelName]?.setStreamHandler(null)
                activeEventChannels.remove(channelName)
                activeStreamSinks.remove(channelName)
            }
            registeredObjects.remove(name)
        }

        /**
         * Check if a name is already registered.
         */
        @JvmStatic
        fun isRegistered(name: String): Boolean = registeredObjects.containsKey(name)

        /**
         * Get all registered class names.
         */
        @JvmStatic
        fun getRegisteredNames(): Set<String> = registeredObjects.keys.toSet()
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        binaryMessenger = binding.binaryMessenger
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)

        // Store plugin instance for late registration
        pluginInstance = this

        // Setup event channels for any already registered stream methods
        setupEventChannels()
    }

    /**
     * Setup EventChannels for all @NativeStream annotated methods.
     */
    private fun setupEventChannels() {
        for ((className, target) in registeredObjects) {
            setupEventChannelsForObject(className, target)
        }
    }

    /**
     * Setup EventChannels for a single registered object.
     * Called when a new object is registered.
     */
    fun setupEventChannelsForObject(className: String, target: Any) {
        val messenger = binaryMessenger ?: return

        val streamMethods = getStreamMethods(target)
        for (methodName in streamMethods) {
            val channelName = "$EVENT_CHANNEL_PREFIX$className.$methodName"

            // Skip if already setup
            if (activeEventChannels.containsKey(channelName)) continue

            val eventChannel = EventChannel(messenger, channelName)
            eventChannel.setStreamHandler(createStreamHandler(className, methodName, target))
            activeEventChannels[channelName] = eventChannel

            android.util.Log.d(TAG, "Setup EventChannel: $channelName")
        }
    }

    /**
     * Create a StreamHandler for a specific stream method.
     */
    private fun createStreamHandler(
        className: String,
        methodName: String,
        target: Any
    ): EventChannel.StreamHandler {
        return object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                if (events == null) return

                val channelName = "$EVENT_CHANNEL_PREFIX$className.$methodName"
                val sinkWrapper = EventSinkWrapper(events)
                activeStreamSinks[channelName] = sinkWrapper

                // Find and invoke the stream method with the sink
                try {
                    val method = findStreamMethod(target, methodName)
                    if (method != null) {
                        invokeStreamMethod(target, method, sinkWrapper, arguments)
                    } else {
                        events.error("NOT_FOUND", "Stream method '$methodName' not found", null)
                    }
                } catch (e: Exception) {
                    events.error("INVOKE_ERROR", e.message ?: "Unknown error", e.stackTraceToString())
                }
            }

            override fun onCancel(arguments: Any?) {
                val channelName = "$EVENT_CHANNEL_PREFIX$className.$methodName"
                activeStreamSinks.remove(channelName)
                android.util.Log.d(TAG, "Stream cancelled: $channelName")
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        // Handle discovery methods
        when (call.method) {
            "_getRegisteredClasses" -> {
                result.success(registeredObjects.keys.toList())
                return
            }
            "_getMethods" -> {
                handleGetMethods(call, result)
                return
            }
            "_getStreams" -> {
                handleGetStreams(call, result)
                return
            }
            "_discover" -> {
                result.success(registeredObjects.mapValues { (_, obj) -> getExposedMethods(obj) })
                return
            }
            "_discoverStreams" -> {
                result.success(registeredObjects.mapValues { (_, obj) -> getStreamMethods(obj) })
                return
            }
        }

        // Handle regular method calls (ClassName.methodName)
        val parts = call.method.split(".")
        if (parts.size != 2) {
            result.error("INVALID_FORMAT", "Method must be 'ClassName.methodName'", null)
            return
        }

        val (className, methodName) = parts
        val target = registeredObjects[className]

        if (target == null) {
            result.error("NOT_FOUND", "Class '$className' not registered", null)
            return
        }

        val method = findMethod(target, methodName)
        if (method == null) {
            result.error("NOT_FOUND", "Method '$methodName' not found or not exposed", null)
            return
        }

        try {
            val returnValue = invokeMethod(target, method, call.arguments)
            result.success(returnValue)
        } catch (e: Exception) {
            result.error("INVOKE_ERROR", e.message ?: "Unknown error", e.stackTraceToString())
        }
    }

    private fun handleGetMethods(call: MethodCall, result: Result) {
        val className = call.arguments as? String
        if (className == null) {
            result.error("INVALID_ARG", "Class name required", null)
            return
        }
        val target = registeredObjects[className]
        if (target == null) {
            result.error("NOT_FOUND", "Class '$className' not registered", null)
            return
        }
        result.success(getExposedMethods(target))
    }

    private fun handleGetStreams(call: MethodCall, result: Result) {
        val className = call.arguments as? String
        if (className == null) {
            result.error("INVALID_ARG", "Class name required", null)
            return
        }
        val target = registeredObjects[className]
        if (target == null) {
            result.error("NOT_FOUND", "Class '$className' not registered", null)
            return
        }
        result.success(getStreamMethods(target))
    }

    private fun findMethod(target: Any, methodName: String): Method? {
        val clazz = target::class.java
        val hasNativeBridge = clazz.isAnnotationPresent(NativeBridge::class.java)

        return clazz.methods.find { method ->
            method.name == methodName && when {
                hasNativeBridge -> {
                    !method.isAnnotationPresent(NativeIgnore::class.java) &&
                    !method.isAnnotationPresent(NativeStream::class.java)
                }
                else -> method.isAnnotationPresent(NativeFunction::class.java)
            }
        }
    }

    private fun getExposedMethods(target: Any): List<String> {
        val clazz = target::class.java
        val hasNativeBridge = clazz.isAnnotationPresent(NativeBridge::class.java)

        return clazz.methods
            .filter { method ->
                // Exclude stream methods from regular methods
                if (method.isAnnotationPresent(NativeStream::class.java)) return@filter false

                when {
                    hasNativeBridge -> !method.isAnnotationPresent(NativeIgnore::class.java)
                    else -> method.isAnnotationPresent(NativeFunction::class.java)
                }
            }
            .filter { it.declaringClass != Any::class.java }
            .map { it.name }
            .distinct()
    }

    /**
     * Get all @NativeStream annotated methods for a target.
     */
    private fun getStreamMethods(target: Any): List<String> {
        val clazz = target::class.java
        return clazz.methods
            .filter { it.isAnnotationPresent(NativeStream::class.java) }
            .filter { it.declaringClass != Any::class.java }
            .map { it.name }
            .distinct()
    }

    /**
     * Find a @NativeStream annotated method by name.
     */
    private fun findStreamMethod(target: Any, methodName: String): Method? {
        val clazz = target::class.java
        return clazz.methods.find { method ->
            method.name == methodName && method.isAnnotationPresent(NativeStream::class.java)
        }
    }

    /**
     * Invoke a stream method, passing the StreamSink as parameter.
     */
    private fun invokeStreamMethod(target: Any, method: Method, sink: StreamSink, args: Any?) {
        method.isAccessible = true

        // Check if method expects StreamSink as first parameter
        val params = method.parameters
        when {
            params.isEmpty() -> {
                // No parameters - shouldn't happen for stream methods
                method.invoke(target)
            }
            params.size == 1 && params[0].type == StreamSink::class.java -> {
                // Just StreamSink
                method.invoke(target, sink)
            }
            params.size == 2 && params[0].type == StreamSink::class.java -> {
                // StreamSink + arguments
                method.invoke(target, sink, args)
            }
            params[0].type == StreamSink::class.java && args is Map<*, *> -> {
                // StreamSink + multiple arguments as map
                val values = mutableListOf<Any?>(sink)
                params.drop(1).forEach { param ->
                    values.add(args[param.name] ?: args[param.name.removePrefix("arg")])
                }
                method.invoke(target, *values.toTypedArray())
            }
            else -> {
                // Fallback: try to invoke with sink and args
                method.invoke(target, sink, args)
            }
        }
    }

    private fun invokeMethod(target: Any, method: Method, args: Any?): Any? {
        method.isAccessible = true

        return when {
            method.parameterCount == 0 -> method.invoke(target)
            method.parameterCount == 1 -> method.invoke(target, args)
            args is Map<*, *> -> {
                val values = method.parameters.map { param ->
                    args[param.name] ?: args[param.name.removePrefix("arg")]
                }.toTypedArray()
                method.invoke(target, *values)
            }
            args is List<*> -> method.invoke(target, *args.toTypedArray())
            else -> method.invoke(target, args)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)

        // Clean up all event channels
        activeEventChannels.values.forEach { it.setStreamHandler(null) }
        activeEventChannels.clear()
        activeStreamSinks.clear()
        binaryMessenger = null
        pluginInstance = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

/**
 * Wraps Flutter's EventChannel.EventSink to implement our StreamSink interface.
 * This allows native code to emit events without depending on Flutter classes directly.
 */
internal class EventSinkWrapper(private val eventSink: EventChannel.EventSink) : StreamSink {

    override fun success(event: Any?) {
        eventSink.success(event)
    }

    override fun error(code: String, message: String?, details: Any?) {
        eventSink.error(code, message, details)
    }

    override fun endOfStream() {
        eventSink.endOfStream()
    }
}

File: ios/Classes/FlutterNativeBridgePlugin.swift
import Flutter
import UIKit

/// Flutter Native Bridge Plugin for iOS
///
/// Bridges Flutter and native iOS code using Objective-C runtime.
/// Supports both MethodChannel (request-response) and EventChannel (streams).
@objc(FlutterNativeBridgePlugin)
public class FlutterNativeBridgePlugin: NSObject, FlutterPlugin {

    private static var registeredObjects: [String: AnyObject] = [:]
    private static var registrar: FlutterPluginRegistrar?

    // Track active event channels and their sinks
    private static var activeEventChannels: [String: FlutterEventChannel] = [:]
    static var activeStreamSinks: [String: EventSinkWrapper] = [:]
    private static var streamHandlers: [String: StreamHandler] = [:]

    static let eventChannelPrefix = "flutter_native_bridge/events/"

    public static func register(with registrar: FlutterPluginRegistrar) {
        self.registrar = registrar

        let channel = FlutterMethodChannel(
            name: "flutter_native_bridge",
            binaryMessenger: registrar.messenger()
        )
        let instance = FlutterNativeBridgePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        // Setup event channels for all registered stream methods
        setupEventChannels()
    }

    /// Setup EventChannels for all stream methods in registered objects.
    private static func setupEventChannels() {
        guard let registrar = registrar else { return }

        for (className, target) in registeredObjects {
            let streamMethods = getStreamMethods(target: target)
            for methodName in streamMethods {
                let channelName = "\(eventChannelPrefix)\(className).\(methodName)"

                // Skip if already setup
                if activeEventChannels[channelName] != nil { continue }

                let eventChannel = FlutterEventChannel(
                    name: channelName,
                    binaryMessenger: registrar.messenger()
                )

                let handler = StreamHandler(className: className, methodName: methodName, target: target)
                eventChannel.setStreamHandler(handler)

                activeEventChannels[channelName] = eventChannel
                streamHandlers[channelName] = handler

                NSLog("FlutterNativeBridge: Setup EventChannel: \(channelName)")
            }
        }
    }

    // MARK: - Registration Methods (Called from FlutterNativeBridge helper)

    static func registerObject(name: String, object: AnyObject) {
        registeredObjects[name] = object
        // Setup event channels for any stream methods in this object
        setupEventChannels()
    }

    static func registerObject(_ object: AnyObject) {
        let name = String(describing: type(of: object))
            .components(separatedBy: ".").last ?? String(describing: type(of: object))
        registeredObjects[name] = object
        // Setup event channels for any stream methods in this object
        setupEventChannels()
    }

    static func unregisterObject(name: String) {
        // Clean up any event channels for this object
        let prefix = "\(eventChannelPrefix)\(name)."
        for channelName in activeEventChannels.keys where channelName.hasPrefix(prefix) {
            activeEventChannels[channelName]?.setStreamHandler(nil)
            activeEventChannels.removeValue(forKey: channelName)
            activeStreamSinks.removeValue(forKey: channelName)
            streamHandlers.removeValue(forKey: channelName)
        }
        registeredObjects.removeValue(forKey: name)
    }

    static func isRegistered(name: String) -> Bool {
        return registeredObjects[name] != nil
    }

    // MARK: - Method Call Handler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // Discovery methods
        switch call.method {
        case "_getRegisteredClasses":
            result(Array(FlutterNativeBridgePlugin.registeredObjects.keys))
            return

        case "_getMethods":
            guard let className = call.arguments as? String else {
                result(FlutterError(code: "INVALID_ARG", message: "Class name required", details: nil))
                return
            }
            guard let target = FlutterNativeBridgePlugin.registeredObjects[className] else {
                result(FlutterError(code: "NOT_FOUND", message: "Class '\(className)' not registered", details: nil))
                return
            }
            result(getExposedMethods(target: target))
            return

        case "_discover":
            var discovery: [String: [String]] = [:]
            for (name, obj) in FlutterNativeBridgePlugin.registeredObjects {
                discovery[name] = getExposedMethods(target: obj)
            }
            result(discovery)
            return

        case "_getStreams":
            guard let className = call.arguments as? String else {
                result(FlutterError(code: "INVALID_ARG", message: "Class name required", details: nil))
                return
            }
            guard let target = FlutterNativeBridgePlugin.registeredObjects[className] else {
                result(FlutterError(code: "NOT_FOUND", message: "Class '\(className)' not registered", details: nil))
                return
            }
            result(FlutterNativeBridgePlugin.getStreamMethods(target: target))
            return

        case "_discoverStreams":
            var discovery: [String: [String]] = [:]
            for (name, obj) in FlutterNativeBridgePlugin.registeredObjects {
                discovery[name] = FlutterNativeBridgePlugin.getStreamMethods(target: obj)
            }
            result(discovery)
            return

        default:
            break
        }

        // Regular method calls (ClassName.methodName)
        let parts = call.method.components(separatedBy: ".")
        guard parts.count == 2 else {
            result(FlutterError(code: "INVALID_FORMAT", message: "Method must be 'ClassName.methodName'", details: nil))
            return
        }

        let className = parts[0]
        let methodName = parts[1]

        guard let target = FlutterNativeBridgePlugin.registeredObjects[className] else {
            result(FlutterError(code: "NOT_FOUND", message: "Class '\(className)' not registered", details: nil))
            return
        }

        // Find and call the method
        let returnValue = invokeMethod(target: target, methodName: methodName, arguments: call.arguments)

        if let error = returnValue as? FlutterError {
            result(error)
        } else if returnValue == nil || returnValue is NSNull {
            result(nil)
        } else {
            result(returnValue)
        }
    }

    // MARK: - Method Invocation

    private func invokeMethod(target: AnyObject, methodName: String, arguments: Any?) -> Any? {
        // Try to find a matching selector
        let selectors = buildSelectors(methodName: methodName, arguments: arguments)

        for selector in selectors {
            if target.responds(to: selector) {
                return performSelector(target: target, selector: selector, arguments: arguments)
            }
        }
        return FlutterError(
            code: "NOT_FOUND",
            message: "Method '\(methodName)' not found. Make sure it's marked with @objc",
            details: nil
        )
    }

    private func buildSelectors(methodName: String, arguments: Any?) -> [Selector] {
        var selectors: [Selector] = []

        if arguments == nil {
            // No arguments
            selectors.append(Selector(methodName))
        } else if let dict = arguments as? [String: Any] {
            // Dictionary arguments - try with parameter names
            let keys = dict.keys.sorted()
            if keys.count == 1 {
                selectors.append(Selector("\(methodName)With\(keys[0].capitalized):"))
                selectors.append(Selector("\(methodName):"))
            } else {
                // Multiple parameters
                var selectorName = methodName
                for (i, key) in keys.enumerated() {
                    if i == 0 {
                        selectorName += "With\(key.capitalized):"
                    } else {
                        selectorName += "\(key):"
                    }
                }
                selectors.append(Selector(selectorName))
                selectors.append(Selector("\(methodName):"))
            }
        } else {
            // Single argument
            selectors.append(Selector("\(methodName):"))
            selectors.append(Selector("\(methodName)With:"))
        }

        return selectors
    }

    private func performSelector(target: AnyObject, selector: Selector, arguments: Any?) -> Any? {
        // Get method return type to handle primitives correctly
        guard let method = class_getInstanceMethod(type(of: target), selector) else {
            return nil
        }

        let returnType = method_copyReturnType(method)
        let fullReturnType = String(cString: returnType)
        free(returnType)

        // Get just the first character which is the actual type encoding
        let returnTypeString = String(fullReturnType.prefix(1))

        // Type encodings: v=void, @=object, c/B=bool, i/s/l/q=int, f/d=float/double

        // Handle void and primitive return types with IMP-based calling
        if returnTypeString != "@" {
            return performPrimitiveSelector(target: target, selector: selector, returnType: returnTypeString, arguments: arguments)
        }

        // Object return type - safe to use perform/takeUnretainedValue
        if arguments == nil {
            let result = target.perform(selector)
            return unwrapResult(result?.takeUnretainedValue())
        }

        // Handle dictionary with single value
        if let dict = arguments as? [String: Any], dict.count == 1, let value = dict.values.first {
            let result = target.perform(selector, with: value)
            return unwrapResult(result?.takeUnretainedValue())
        }

        // Pass arguments directly
        let result = target.perform(selector, with: arguments)
        return unwrapResult(result?.takeUnretainedValue())
    }

    private func performPrimitiveSelector(target: AnyObject, selector: Selector, returnType: String, arguments: Any?) -> Any? {
        // Get IMP for direct calling
        guard let imp = class_getMethodImplementation(type(of: target), selector) else {
            return nil
        }

        switch returnType {
        case "v": // void
            typealias VoidIMP = @convention(c) (AnyObject, Selector) -> Void
            let fn = unsafeBitCast(imp, to: VoidIMP.self)
            fn(target, selector)
            return nil

        case "c", "B": // char/Bool
            typealias BoolIMP = @convention(c) (AnyObject, Selector) -> Bool
            let fn = unsafeBitCast(imp, to: BoolIMP.self)
            return fn(target, selector)

        case "i", "s", "l", "q": // signed int types
            typealias IntIMP = @convention(c) (AnyObject, Selector) -> Int
            let fn = unsafeBitCast(imp, to: IntIMP.self)
            return fn(target, selector)

        case "I", "S", "L", "Q": // unsigned int types
            typealias UIntIMP = @convention(c) (AnyObject, Selector) -> UInt
            let fn = unsafeBitCast(imp, to: UIntIMP.self)
            return Int(fn(target, selector))

        case "f": // float
            typealias FloatIMP = @convention(c) (AnyObject, Selector) -> Float
            let fn = unsafeBitCast(imp, to: FloatIMP.self)
            return Double(fn(target, selector))

        case "d": // double
            typealias DoubleIMP = @convention(c) (AnyObject, Selector) -> Double
            let fn = unsafeBitCast(imp, to: DoubleIMP.self)
            return fn(target, selector)

        default:
            return nil
        }
    }

    private func unwrapResult(_ value: Any?) -> Any? {
        guard let value = value else { return nil }

        // Convert NSNumber to appropriate Swift types
        if let number = value as? NSNumber {
            let type = String(cString: number.objCType)
            switch type {
            case "c", "B": return number.boolValue
            case "i", "s", "l", "q": return number.intValue
            case "f", "d": return number.doubleValue
            default: return number
            }
        }

        return value
    }

    // MARK: - Method Discovery

    private func getExposedMethods(target: AnyObject) -> [String] {
        var methods: Set<String> = []
        let streamMethods = FlutterNativeBridgePlugin.getStreamMethods(target: target)

        var methodCount: UInt32 = 0
        guard let methodList = class_copyMethodList(type(of: target), &methodCount) else {
            return []
        }

        for i in 0..<Int(methodCount) {
            let method = methodList[i]
            let selector = method_getName(method)
            var name = NSStringFromSelector(selector)

            // Filter out system methods
            if name.hasPrefix("_") || name.hasPrefix(".") { continue }
            if name.contains("init") || name.contains("dealloc") { continue }
            if name.contains(".cxx") { continue }

            // Extract method name (before first colon)
            if let colonIndex = name.firstIndex(of: ":") {
                name = String(name[..<colonIndex])
            }

            // Remove "With" suffix if present
            if name.hasSuffix("With") {
                name = String(name.dropLast(4))
            }

            // Exclude stream methods from regular methods
            if streamMethods.contains(name) { continue }

            // Only include user-defined methods
            if !name.isEmpty && name.first?.isLowercase == true {
                methods.insert(name)
            }
        }

        free(methodList)
        return Array(methods).sorted()
    }

    // MARK: - Stream Discovery

    /// Get all stream methods for a target object.
    /// Stream methods are identified by having a StreamSink parameter.
    private static func getStreamMethods(target: AnyObject) -> [String] {
        var methods: Set<String> = []

        var methodCount: UInt32 = 0
        guard let methodList = class_copyMethodList(type(of: target), &methodCount) else {
            return []
        }

        for i in 0..<Int(methodCount) {
            let method = methodList[i]
            let selector = method_getName(method)
            let selectorName = NSStringFromSelector(selector)

            // Filter out system methods
            if selectorName.hasPrefix("_") || selectorName.hasPrefix(".") { continue }
            if selectorName.contains("init") || selectorName.contains("dealloc") { continue }
            if selectorName.contains(".cxx") { continue }

            // Check if this method has a StreamSink parameter
            // Methods with StreamSink will have selector like "methodNameWithSink:" or "methodName:"
            // We check the method signature for StreamSink type
            let numberOfArgs = method_getNumberOfArguments(method)

            // Check each argument type (skip self and _cmd which are first 2)
            var hasStreamSink = false
            for argIndex in 2..<numberOfArgs {
                if let argType = method_copyArgumentType(method, argIndex) {
                    let typeString = String(cString: argType)
                    free(argType)

                    // Check if argument is an object that conforms to StreamSink
                    // The type encoding for id<StreamSink> or similar will contain "@"
                    if typeString.contains("@") {
                        // We need to check if the selector contains "sink" (case insensitive)
                        // This is a heuristic since we can't directly check protocol conformance at runtime
                        let lowerSelector = selectorName.lowercased()
                        if lowerSelector.contains("sink") {
                            hasStreamSink = true
                            break
                        }
                    }
                }
            }

            if hasStreamSink {
                var name = selectorName

                // Extract method name (before first colon)
                if let colonIndex = name.firstIndex(of: ":") {
                    name = String(name[..<colonIndex])
                }

                // Remove "WithSink" suffix if present
                if name.hasSuffix("WithSink") {
                    name = String(name.dropLast(8))
                }

                if !name.isEmpty && name.first?.isLowercase == true {
                    methods.insert(name)
                }
            }
        }

        free(methodList)
        return Array(methods).sorted()
    }

    /// Find and invoke a stream method on a target.
    static func invokeStreamMethod(target: AnyObject, methodName: String, sink: StreamSink, arguments: Any?) {
        // Build possible selectors for stream methods
        let selectors = [
            Selector("\(methodName)WithSink:"),
            Selector("\(methodName):"),
            Selector("\(methodName)WithSink:arguments:"),
            Selector("\(methodName):arguments:")
        ]

        for selector in selectors {
            if target.responds(to: selector) {
                if selector.description.contains("arguments") {
                    _ = target.perform(selector, with: sink, with: arguments)
                } else {
                    _ = target.perform(selector, with: sink)
                }
                return
            }
        }

        // Method not found - send error
        sink.error(code: "NOT_FOUND", message: "Stream method '\(methodName)' not found", details: nil)
    }
}

// MARK: - Stream Handler

/// Handles FlutterEventChannel stream lifecycle.
private class StreamHandler: NSObject, FlutterStreamHandler {
    let className: String
    let methodName: String
    weak var target: AnyObject?

    init(className: String, methodName: String, target: AnyObject) {
        self.className = className
        self.methodName = methodName
        self.target = target
        super.init()
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        guard let target = target else {
            return FlutterError(code: "TARGET_GONE", message: "Target object no longer exists", details: nil)
        }

        let channelName = "\(FlutterNativeBridgePlugin.eventChannelPrefix)\(className).\(methodName)"
        let sinkWrapper = EventSinkWrapper(eventSink: events)
        FlutterNativeBridgePlugin.activeStreamSinks[channelName] = sinkWrapper

        // Invoke the stream method with the sink
        FlutterNativeBridgePlugin.invokeStreamMethod(
            target: target,
            methodName: methodName,
            sink: sinkWrapper,
            arguments: arguments
        )

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        let channelName = "\(FlutterNativeBridgePlugin.eventChannelPrefix)\(className).\(methodName)"
        FlutterNativeBridgePlugin.activeStreamSinks.removeValue(forKey: channelName)
        return nil
    }
}

// MARK: - Event Sink Wrapper

/// Wraps Flutter's EventSink to implement our StreamSink protocol.
class EventSinkWrapper: NSObject, StreamSink {
    private let eventSink: FlutterEventSink

    init(eventSink: @escaping FlutterEventSink) {
        self.eventSink = eventSink
        super.init()
    }

    func success(_ event: Any?) {
        eventSink(event)
    }

    func error(code: String, message: String?, details: Any?) {
        eventSink(FlutterError(code: code, message: message, details: details))
    }

    func endOfStream() {
        eventSink(FlutterEndOfEventStream)
    }
}
